README.md

```diff
@@ -0,0 +1,237 @@
+
+### End result / usage
+
+The classes in this package allow for the following API and usage:
+
+```java
+// Building a simple container object not linked to any test framework
+final Container container = Container.builder()
+  .withImage("alpine:3.5")
+  .withExposedPort(1234)
+  .build();
+
+// now that the container is built, the `with___` methods disappear and runtime methods,
+//  like `exec` become available
+container.start();
+container.exec("date");
+container.stop();
+```
+
+
+
+### Implementation
+
+* `Container` holds configuration data in fields, plus methods for 'run-time' interactions such as `start`, `stop` , `exec`, etc.
+* `ContainerBuilder` (and its parent class `BaseContainerBuilder`)  exposes the define-time `withXYZ` fluent setter API
+* `BaseContainerBuilder` uses package-scoped access to `Container`'s configuration fields to apply values before `build()` is called
+* `ContainerBuilder` can also validate the container's configuration just prior to building, e.g. to ensure that essential fields like `image` are set
+
+
+
+
+---
+
+## Test-framework agnostic vs Test-framework coupled classes
+
+The above example shows test-framework agnostic usage. Still to support use of JUnit 4 rules, we need an alternative Container implementation (implementing JUnit's `TestRule`) and an alternative Builder specialisation. 
+
+### End result / usage (JUnit 4 case)
+
+See the package `org.testcontainers.testing.junit4` - **this would be published as a Testcontainers JAR**.
+
+The end result achieved is almost the same as the existing Testcontainers 1.x API. The only differences are usage of a different class name and use of the builder pattern.
+
+```java
+public class SimpleJUnit4Test {
+
+    @Rule
+    public ContainerTestRule alpine = ContainerTestRule.builder()
+            .withImage("alpine:3.5")
+            .withExposedPort(1234)
+            .build();
+  
+    @Test
+    public void doNothingTest() {
+        alpine.exec("true");
+    }
+
+```
+
+
+
+### Implementation
+
+* `ContainerTestRule` should have a run-time API that matches that of `Container`. To achieve this with least fuss, we can use Lombok's `@Delegate`. In using Lombok, though, we should be certain to de-Lombok all published sources.
+* `ContainerTestRuleBuilder` uses the common `BaseContainerBuilder` class so that the same fluent setter API is available when defining containers. A small but encapsulated amount of generics trickery is needed so that the fluent setter API returns the expected 'self' type.
+
+
+
+### End result / usage (TestNG case; potentially others)
+
+See `org.testcontainers.testing.testng`. *This is not mature, possibly not idiomatic, and only reflects a draft idea:*
+
+```java
+public class SimpleTestngTest extends TestcontainersBaseTest {
+
+    private Container alpine = Container.builder()
+            .withImage("alpine:3.5")
+            .build();
+
+    @Test
+    public void doNothing() {
+        alpine.exec("date");
```

Comment:
re `exec` and others. Since the number of such "helpers" will grow, how about introducing a concept of "recipes"? i.e. `ExecRecipe`, `GetFileRecipe`, etc?

Reply:
I think in general we want a OO abstraction of docker-java, so I assume we will have all basic docker API methods available, like:
  * cp
  * exec
  * diff
  * ...

I always wonder how much to deviate from the known Docker API or if it would be kind of enough to conveniently expose these methods.

Reply:
I'm not sure that this is the goal actually :D The goal is to have a high-level API to start and manipulate the containers, and cp/exec/diff/etc are too low level for that

Reply:
Yes, this is a good thing to discuss. The way I see it, the library should be 'narrow but deep':

Narrow: we don't need to cover every possible method of the docker API (probably >50%, but definitely <100%)
Deep: we should aim to do as much as possible for the user to hide the 'accidental complexity' of docker - e.g. all the things around file mounting, copying, and network port mapping. 

By hiding some of this complexity we do reduce the flexibility of the docker API, but this is unavoidable.

There might be people who want a 100% wide abstraction, but TBH I think these people _need_ the finer-grained, lower-level tools that docker-java provides, so we're OK to avoid this area.

Reply:
@bsideup I don't think `cp` and `exec` are too low level per se, since both of your examples are basically renamed calls to these methods, aren't they ? ;)

There might be more convenient API composed on top of the basic APIs, but I don't see any pain in exposing them in a public API, still more convenient to use in a OO manner then docker-java.